### PR TITLE
Fix unit test

### PR DESCRIPTION
**Pull request by allaffa**
The existing test builds a graph with a body-centered-cubic (BCC) structure and assigns random nodal features

Changes: 
- `test_random_graph.py` removed
- `test_deterninistic_graph.py` added
- `utils/random_graphs.py` removed
- `utils/deterministic_graphs.py` added
- thresholding for assert to pass the test is different depending on the model used. GIN and GAT are renown to have a worse performance that PIN and MFC. Even for small graphs, their performance is not as good as the other twos. So I had to use two different thresholds

### CODE REVIEW

File: examples/ci.json [hunk 1]
```diff
@@ -17,10 +17,10 @@
     "NeuralNetwork": {
         "Architecture": {
             "model_type": "PNN",
-            "radius": 7,
+            "radius": 4,
```
Comment (streeve) [inline examples/ci.json:20]: Is this size reduction mainly to save time?

File: test_graphs.py [hunk 2]
```diff
@@ -2,18 +2,22 @@
...
 
 from run_config_input import run_normal_config_file
-from utils.random_graph_data import random_graph_data
+from utils.deterministic_graph_data import deterministic_graph_data
 from utils.utils import get_comm_size_and_rank
 from test_trained_model import test_trained_model
 
+import torch
+
+torch.manual_seed(0)
```
Comment (streeve) [inline test_graphs.py:11]: Why is this here?

File: utils/deterministic_graph_data.py [hunk 3]
```diff
@@ -40,22 +45,33 @@ def random_graph_data(
...
+    number_atoms = 2 * number_unit_cell_x * number_unit_cell_y * number_unit_cell_z
+    positions = torch.zeros(number_atoms, 3)
+    lattice_parameter = 0.5
+
+    # We assume that the unit cell is Body Center Cubic (BCC)
+    count_pos = 0
+    for x in range(0, number_unit_cell_x):
+        for y in range(0, number_unit_cell_y):
+            for z in range(0, number_unit_cell_z):
+                positions[count_pos][0] = x + 0.0
```
Comment (streeve) [inline utils/deterministic_graph_data.py:57]: `+0` not really needed

File: utils/deterministic_graph_data.py [hunk 4]
```diff
@@ -40,22 +45,33 @@ def random_graph_data(
...
+                positions[count_pos][1] = y + 0.0
+                positions[count_pos][2] = z + 0.0
+                positions[count_pos + 1][0] = x + lattice_parameter
```
Comment (streeve) [inline utils/deterministic_graph_data.py:60]: Should be + 1/2 the lattice parameter for BCC

File: test_graphs.py [hunk 5]
```diff
@@ -2,18 +2,22 @@
...
+
 
 @pytest.mark.parametrize("model_type", ["GIN", "GAT", "MFC", "PNN"])
```
Comment (streeve) [inline test_graphs.py:14]: MFC fails on the assert if run first - I don't understand why, but that's concerning

File: test_graphs.py [hunk 6]
```diff
@@ -33,4 +37,7 @@ def pytest_train_model(model_type):
     error, true_values, predicted_values = test_trained_model(tmp_file, model_type)
 
     for true_value, predicted_value in zip(true_values, predicted_values):
-        assert abs(true_value[0] - predicted_value[0]) < 1e-3
+        if model_type == "PIN" or model_type == "MFC":
+            assert abs(true_value[0] - predicted_value[0]) < 0.1
```
Comment (streeve) [inline test_graphs.py:41]: Based on running these locally I think we could tighten these tolerances a little (but until it's clear why MFC fails when run separately I'm not sure)

File: examples/ci.json (hunk 1 above)
Comment (allaffa) [reply examples/ci.json:20]: Yes

File: test_graphs.py (hunk 2 above)
Comment (allaffa) [reply test_graphs.py:11]: I thought that fixing the random seed would help ensure that the unit_tests run somewhat consistently on different machines. Do you recommend thatI remove it?

File: utils/deterministic_graph_data.py (hunk 4 above)
Comment (allaffa) [reply utils/deterministic_graph_data.py:60]: I already use 0.5. I set `lattice_parameter = 0.5` a few lines above. 
The unit cell has 0.5 as standard lattice size. Then this is pre-multiplied by the lattice parameter that attains the minimum energy configuration for each material. But in this case I would keep the standard lattice size of the unit cell, which is 0.5. I already use 0.5 for the unit test.

Comment (streeve) [reply utils/deterministic_graph_data.py:60]: Sorry, I misread. I would just use `+0.5` directly to avoid confusion - eventually we can rewrite this to use any lattice basis

File: test_graphs.py (hunk 2 above)
Comment (streeve) [reply test_graphs.py:11]: I just wasn't familiar with the function and was confused why it wasn't inside `pytest_train_model`

File: test_graphs.py (hunk 5 above)
Comment (pzhanggit) [reply test_graphs.py:14]: MFC fails in my local runs. Any difference between running locally and here?

Comment (streeve) [reply test_graphs.py:14]: Hopefully just random fluctuations then - I get `0.1134`. Maybe we should just use different tolerances for each

File: test_graphs.py [hunk 7]
```diff
@@ -33,4 +37,7 @@ def pytest_train_model(model_type):
...
+        else:
+            assert abs(true_value[0] - predicted_value[0]) < 0.5
```
Comment (pzhanggit) [inline test_graphs.py:43]: Is there a reason that different error thresholds are used for different models?

**Review by pzhanggit — COMMENTED** [review]
The major concern I have is about the model accuracy.  
1. Is the error at 0.1 or 0.5 considered small in the unit test?
2. Are PIN and MFC expected to have better accuracy than the other two models?

File: test_graphs.py (hunk 6 above)
Comment (allaffa) [reply test_graphs.py:41]: The problem is as small as it can be. The structure has 4 atoms. Smaller than this makes no sense. Bigger than this, the structure becomes combinatorially complex, so the problem becomes fo a comparable challenge as the DFT datasets. 
Should we relax the tolerance to 0.5 for all the models?
Otherwise, I have no immediate solutions to the task of developing the unit test.

File: test_graphs.py (hunk 7 above)
Comment (allaffa) [reply test_graphs.py:43]: See comments at the top of the description of the PR

File: test_graphs.py (hunk 6 above)
Comment (allaffa) [reply test_graphs.py:41]: GIN, GAT, and MFC use aggregations that can ambiguously identify two different graphs as equivalent. 
PIN mixes and simultaneously uses the aggregations used by the previous thee models to make sure that different graphs are properly recognized as different. This is one way to explain why PIN does not show up in any of the inconsistencies we are dealing with now.

Comment (streeve) [reply test_graphs.py:41]: I was only thinking of test accuracy here (same as Pei), not test size. Based on running locally, I think we could use 0.05 for PNN, 0.2 for MFC, and 0.4 for GIN/GAT. But maybe I'm trying to be too aggressive.

File: test_graphs.py [hunk 8]
```diff
@@ -33,4 +37,7 @@ def pytest_train_model(model_type):
     error, true_values, predicted_values = test_trained_model(tmp_file, model_type)
 
     for true_value, predicted_value in zip(true_values, predicted_values):
-        assert abs(true_value[0] - predicted_value[0]) < 1e-3
+        if model_type == "PIN" or model_type == "MFC":
```
Comment (streeve) [inline test_graphs.py:40]: PIN -> PNN

File: test_graphs.py (hunk 6 above)
Comment (streeve) [reply test_graphs.py:41]: With your last comment, we could just use 0.1 for PNN and 0.4 for the rest and try to avoid unnecessary failures

**Comment by allaffa** [comment]
@streeve @pzhanggit 
I applied your suggestions. Let me know if this solves the local issues you are facing.

**Comment by allaffa** [comment]
@pzhanggit @streeve 
Things are not looking good yet. I need to work more on this :)
